controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitTransactionTask.java

```diff
@@ -0,0 +1,174 @@
+            } else {
+                log.debug("Successfully committed transactions on epoch {} on stream {}/{}", epoch, scope, stream);
+                if (processedEvents != null) {
+                    processedEvents.offer(event);
+                }
+            }
+        }, executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> writeBack(CommitEvent event) {
+        return streamMetadataTasks.writeEvent(event);
+    }
+
+    private CompletableFuture<Void> tryCommitTransactions(final String scope,
+                                                  final String stream,
+                                                  final int epoch,
+                                                  final OperationContext context,
+                                                  final String delegationToken) {
+        // if node already exists and doesnt match event.transactionId, throw operation not allowed. dont worry,
+        // it will be posted back in the stream and retried later. Generally if a transaction commit starts, it will come to
+        // an end.. but during failover,
+        // once we have created the node, we are guaranteed that it will be only that transaction that will be getting
+        // committed at that time.. so post failover even when we recover, we will handle one transaction at a time.
+        CompletableFuture<List<UUID>> txnListFuture = streamMetadataStore.getTxnCommitList(scope, stream, context, executor)
+                .thenCompose(record -> {
+                    if (record == null) {
+                        // no ongoing commits on transactions.
+                        return createNewTxnCommitList(scope, stream, epoch, context, executor);
+                    } else {
+                        // check if the epoch in record matches current epoch. if not throw OperationNotAllowed
+                        if (record.getEpoch() == epoch) {
+                            // Note: If there are transactions that are not included in the commitList but have committing state,
+                            // we can be sure they will be completed through another event for this epoch.
+                            return CompletableFuture.completedFuture(record.getTransactionsToCommit());
+                        } else {
+                            log.debug("Postponing commit on epoch {} as transactions on different epoch {} are being committed for stream {}/{}",
+                                    epoch, record.getEpoch(), scope, stream);
+                            throw StoreException.create(StoreException.Type.OPERATION_NOT_ALLOWED,
+                                    "Transactions on different epoch are being committed");
+                        }
+                    }
+                });
+
+        CompletableFuture<Void> commitFuture = txnListFuture.thenCompose(transactionsToCommit -> streamMetadataStore.getActiveSegmentIds(scope, stream, epoch, context, executor)
```

Comment:
thenComposeAsync, everywhere in this class

Reply:
do you mean to suggest to always use thenComposeAsync? 
using `thenCompose` will have a better chance of being executed on the same thread where the original future completed. 
From CompletableFuture documentation: 
```Actions supplied for dependent completions of non-async methods may be performed by the thread that completes the current CompletableFuture, or by any other caller of a completion method.```
So using thenCompose

Reply:
`thenCompose` is fine too, but you need to be careful where those previous tasks execute. If they run on a different executor, you may want to use `async` to transfer callbacks to a different executor. I only wanted you to be aware of this; but if you have the same executor throughout, it's fine.

You might also want to run the Controller and do a thread dump at one point and see if the ForkJoinPool is used for anything. Ideally it should not.